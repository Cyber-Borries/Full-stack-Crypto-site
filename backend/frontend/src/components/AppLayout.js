import React, { useEffect, useState } from "react";
import "react-toastify/dist/ReactToastify.css";
import Table from "react-bootstrap/Table";
import "bootstrap/dist/css/bootstrap.min.css";
import Card from "react-bootstrap/Card";
import "./appLayoutCSS.css";
import HashLoader from "react-spinners/HashLoader";
import CircleChart from "./CircleChart";

export default function AppLayout(props) {
  //state to capture amount of ZAR to convert
  const [chart, setChart] = useState([]);

  const [selectedCoin, setSelectedCoin] = useState();
  const [isLoading, setIsLoading] = useState(true);
  let [color, setColor] = useState("#0ff");
  //added an empty array to handle the api response from the backend and save the coins
  const [coins, setCoins] = useState();

  useEffect(() => {
    // Fetch API data
    fetch("/api")
      .then((response) => response.json())
      .then((data) => {
        setCoins(data);
        if (data.length !== 0) {
          let newArr = coins;
          let newCopy = newArr.splice(0, 10);
          setChart(newCopy);
        }
        setIsLoading(false);
      });
  }, [coins]);

  const handleChange = (event) => {
    setSelectedCoin(event.target.value);
  };

  let selectedData;

  if (selectedCoin) {
    selectedData = coins.find((item) => item.name === selectedCoin);
  }

  return (
    <>
      <div>
        {isLoading ? (
          <div className="loader">
            <div className="ball"> </div>{" "}
            <HashLoader
              color={color}
              loading={isLoading}
              size={350}
              aria-label="Loading Spinner"
              data-testid="loader"
            />
            <h4 className="warning">Warning:</h4>
            <h6 className="loading-text">
              Crypto is volatile. Do your research!
            </h6>
          </div>
        ) : (
          <>
            <div className="chart">
              <CircleChart chart={chart} />
              <h1>Circle chart showing market dominance of the top 10 coins</h1>
            </div>
            <div className="crypto-table">
              <h1>
                Below is a list of the top 100 coins as per Coinmarketcap
                &trade;
              </h1>

              <Table>
                <tbody>
                  <tr className="table-headings">
                    <td>Coin name</td>

                    <td>Coin price in $</td>
                    <td>Coin price in ZAR</td>
                    <td>Coin market cap</td>
                    <td>1 hour change</td>
                    <td>24 hour change</td>
                    <td>7 day change</td>
                    <td>Market cap dominance</td>
                  </tr>
                  {coins?.map((coin) => (
                    <>
                      <tr>
                        <td>{coin.name}</td>

                        <td>
                          {Intl.NumberFormat("en-ZA", {
                            style: "currency",
                            currency: "USD",
                          }).format(coin.quote.USD.price)}
                        </td>
                        <td>
                          {Intl.NumberFormat("en-ZA", {
                            style: "currency",
                            currency: "ZAR",
                          }).format(coin.quote.USD.price * 17.14)}
                        </td>
                        <td>
                          {Intl.NumberFormat("en-ZA", {
                            style: "currency",
                            currency: "USD",
                          }).format(coin.quote.USD.market_cap)}
                        </td>

                        <td
                          className={
                            coin.quote.USD.percent_change_1h > 0
                              ? "class1"
                              : "class2"
                          }
                        >
                          {coin.quote.USD.percent_change_1h.toFixed(2)} %
                        </td>
                        <td
                          className={
                            coin.quote.USD.percent_change_24h > 0
                              ? "class1"
                              : "class2"
                          }
                        >
                          {coin.quote.USD.percent_change_24h.toFixed(2)} %
                        </td>
                        <td
                          className={
                            coin.quote.USD.percent_change_24h > 0
                              ? "class1"
                              : "class2"
                          }
                        >
                          {coin.quote.USD.percent_change_7d.toFixed(2)} %
                        </td>
                        <td>
                          {coin.quote.USD.market_cap_dominance.toFixed(2)} %
                        </td>
                      </tr>
                    </>
                  ))}
                </tbody>
              </Table>
            </div>
            <div className="crypto-info">
              <h2>
                Select a crypto currency from the dropdown list below to see
                more information
              </h2>
              <div className="select">
                <select
                  value={selectedCoin}
                  onChange={(e) => setSelectedCoin(e.target.value)}
                >
                  {coins?.map((coin) => (
                    <>
                      <option key={coin.name} value={coin.name}>
                        {coin.name}
                      </option>
                    </>
                  ))}
                </select>
              </div>
              {selectedData && (
                <div>
                  <Card style={{ width: "18rem" }}>
                    <Card.Body>
                      <Card.Title>
                        <span className="crypto-name">{selectedData.name}</span>
                      </Card.Title>

                      <Card.Title>
                        Coin rank:
                        <br />
                        {selectedData.cmc_rank}
                      </Card.Title>
                      <Card.Title>
                        Date added: <br />
                        {selectedData.date_added.slice(0, 10)}
                      </Card.Title>

                      <Card.Text>
                        Price <br />
                        {Intl.NumberFormat("en-ZA", {
                          style: "currency",
                          currency: "ZAR",
                        }).format(selectedData.quote.USD.price * 17.14)}
                      </Card.Text>
                      <Card.Text>
                        Circulating Supply <br />
                        {selectedData.circulating_supply.toLocaleString(
                          undefined,
                          {
                            maximumFractionDigits: 2,
                          }
                        )}
                      </Card.Text>

                      <Card.Text>
                        Market Cap <br />
                        {selectedData.quote.USD.market_cap.toLocaleString(
                          undefined,
                          {
                            maximumFractionDigits: 2,
                          }
                        )}
                      </Card.Text>
                      <Card.Text
                        className={
                          selectedData.quote.USD.percent_change_1h > 0
                            ? "class1"
                            : "class2"
                        }
                      >
                        % change in the last 1h <br />
                        {selectedData.quote.USD.percent_change_1h.toLocaleString(
                          undefined,
                          { maximumFractionDigits: 2 }
                        )}
                        %
                      </Card.Text>
                      <Card.Text
                        className={
                          selectedData.quote.USD.percent_change_24h > 0
                            ? "class1"
                            : "class2"
                        }
                      >
                        % change in the last 24h <br />
                        {selectedData.quote.USD.percent_change_24h.toLocaleString(
                          undefined,
                          { maximumFractionDigits: 2 }
                        )}
                        %
                      </Card.Text>
                    </Card.Body>
                  </Card>
                </div>
              )}
            </div>
          </>
        )}
      </div>
    </>
  );
}
