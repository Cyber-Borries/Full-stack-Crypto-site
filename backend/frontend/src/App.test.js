import React from "react";
import renderer from "react-test-renderer";
import Button from "react-bootstrap/Button";
import Card from "react-bootstrap/Card";

import AppLayout from "../src/components/AppLayout";

const fakeData = {
  id: 1,
  name: "Bitcoin",
  symbol: "BTC",
  slug: "bitcoin",
  num_market_pairs: 9939,
  date_added: "2013-04-28T00:00:00.000Z",
  tags: [
    "mineable",
    "pow",
    "sha-256",
    "store-of-value",
    "state-channel",
    "coinbase-ventures-portfolio",
    "three-arrows-capital-portfolio",
    "polychain-capital-portfolio",
    "binance-labs-portfolio",
    "blockchain-capital-portfolio",
    "boostvc-portfolio",
    "cms-holdings-portfolio",
    "dcg-portfolio",
    "dragonfly-capital-portfolio",
    "electric-capital-portfolio",
    "fabric-ventures-portfolio",
    "framework-ventures-portfolio",
    "galaxy-digital-portfolio",
    "huobi-capital-portfolio",
    "alameda-research-portfolio",
    "a16z-portfolio",
    "1confirmation-portfolio",
    "winklevoss-capital-portfolio",
    "usv-portfolio",
    "placeholder-ventures-portfolio",
    "pantera-capital-portfolio",
    "multicoin-capital-portfolio",
    "paradigm-portfolio",
  ],
  max_supply: 21000000,
  circulating_supply: 19265981,
  total_supply: 19265981,
  platform: null,
  cmc_rank: 1,
  self_reported_circulating_supply: null,
  self_reported_market_cap: null,
  tvl_ratio: null,
  last_updated: "2023-01-18T19:09:00.000Z",
  quote: {
    USD: {
      price: 20923.215068597874,
      volume_24h: 29552756748.4045,
      volume_change_24h: 16.2217,
      percent_change_1h: 0.03797112,
      percent_change_24h: -1.97406821,
      percent_change_7d: 19.72441483,
      percent_change_30d: 26.20844804,
      percent_change_60d: 25.76825202,
      percent_change_90d: 9.58437787,
      market_cap: 403106263970.5203,
      market_cap_dominance: 41.3915,
      fully_diluted_market_cap: 439387516440.56,
      tvl: null,
      last_updated: "2023-01-18T19:09:00.000Z",
    },
  },
};

test("renders correctly", () => {
  const tree = renderer
    .create(
      <Card style={{ width: "18rem" }}>
        <Card.Body>
          <Card.Title>
            <span className="crypto-name">{fakeData.name}</span>
          </Card.Title>

          <Card.Title>
            Coin rank:
            <br />
            {fakeData.cmc_rank}
          </Card.Title>
          <Card.Title>
            Date added: <br />
            {fakeData.date_added.slice(0, 10)}
          </Card.Title>

          <Card.Text>
            Price <br />
            {Intl.NumberFormat("en-ZA", {
              style: "currency",
              currency: "ZAR",
            }).format(fakeData.quote.USD.price * 17.14)}
          </Card.Text>
          <Card.Text>
            Circulating Supply <br />
            {fakeData.circulating_supply.toLocaleString(undefined, {
              maximumFractionDigits: 2,
            })}
          </Card.Text>

          <Card.Text>
            Market Cap <br />
            {fakeData.quote.USD.market_cap.toLocaleString(undefined, {
              maximumFractionDigits: 2,
            })}
          </Card.Text>
          <Card.Text
            className={
              fakeData.quote.USD.percent_change_1h > 0 ? "class1" : "class2"
            }
          >
            % change in the last 1h <br />
            {fakeData.quote.USD.percent_change_1h.toLocaleString(undefined, {
              maximumFractionDigits: 2,
            })}
            %
          </Card.Text>
          <Card.Text
            className={
              fakeData.quote.USD.percent_change_24h > 0 ? "class1" : "class2"
            }
          >
            % change in the last 24h <br />
            {fakeData.quote.USD.percent_change_24h.toLocaleString(undefined, {
              maximumFractionDigits: 2,
            })}
            %
          </Card.Text>
        </Card.Body>
      </Card>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
