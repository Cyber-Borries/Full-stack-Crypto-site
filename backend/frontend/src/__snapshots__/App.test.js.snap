// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="card"
  style={
    Object {
      "width": "18rem",
    }
  }
>
  <div
    className="card-body"
  >
    <div
      className="card-title h5"
    >
      <span
        className="crypto-name"
      >
        Bitcoin
      </span>
    </div>
    <div
      className="card-title h5"
    >
      Coin rank:
      <br />
      1
    </div>
    <div
      className="card-title h5"
    >
      Date added: 
      <br />
      2013-04-28
    </div>
    <p
      className="card-text"
    >
      Price 
      <br />
      R 358 623,91
    </p>
    <p
      className="card-text"
    >
      Circulating Supply 
      <br />
      19 265 981
    </p>
    <p
      className="card-text"
    >
      Market Cap 
      <br />
      403 106 263 970,52
    </p>
    <p
      className="class1 card-text"
    >
      % change in the last 1h 
      <br />
      0,04
      %
    </p>
    <p
      className="class2 card-text"
    >
      % change in the last 24h 
      <br />
      -1,97
      %
    </p>
  </div>
</div>
`;
